### DISCUSSION TITLE
whats wrong with ":refer :all"?

### DISCUSSION THREAD

**Discussion by martianhi** [clj-kondo/clj-kondo#1797] [Q&A]
editing the first lines of the playground:
```
(ns foo
  (:require
    [clojure.string :refer :all]
```

reports "use alias or :refer". 

because it trips up here, the linter doesn't recognise any functions from the required namespace later in the code, as they aren't prepended with any ns alias or anything. it reports errors for all of them.

i'm a clj novice, i got to using `:refer :all` from an auto-generated compojure skeleton.

here is compojure's hello world example:

```
(ns hello-world.core
  (:require [compojure.core :refer :all]
            [compojure.route :as route]))

(defroutes app
  (GET "/" [] "<h1>Hello World</h1>")
  (route/not-found "<h1>Page not found</h1>"))
```

i figure there's nothing wrong with such a declaration, and the linter is just being overly sensitive?

it's also here:
https://clojuredocs.org/clojure.core/ns#example-542692d4c026201cdc327014

https://guide.clojure.style/#prefer-require-over-use
reports it is "acceptable as warranted"

is it possible to disable this check, or otherwise work around it? i would still like to be able to use clj-kondo to improve my still shoddy code.

**Comment by borkdude** [comment]
It is possible to disable this check, like with any checks in clj-kondo, but it makes linting less reliable since it's less clear where symbols that you are using are coming from. And this is exactly the thing that is wrong with `:refer :all`: as a reader of the code I have to keep in my head the things that came from this namespace, which increases mental overhead.
Also: ClojureScript doesn't even support `:refer :all` so if you want to write portable code, it's best to avoid it.
Just using an alias `[clojure.string :as str]` and then typing `str/...` will also make auto-completion work more reliable.

In summary:

- `:refer :all` increases mental overhead
- makes static analysis harder an tooling (clj-kondo, auto-completion)
- isn't portable across Clojure dialects

**Comment by martianhi** [comment]
thanks for the reply. it's a tiny project so i'm happy to disable it. i don't mind a warning about it, but its unfortunate that it necessarily creates errors for all other functions.

**Reply by borkdude** [reply]
@mooseyboots Let me upgrade clj-kondo on the playground, it should not create any errors when you have just one :refer :all.

**Reply by borkdude** [reply]
Hmm, I'm not sure what you mean with "creates errors for all other functions".

This works like expected, I think:

![Screenshot 2022-09-08 at 15 14 14](https://user-images.githubusercontent.com/284934/189131293-0f199f51-3d52-499e-a985-1be1e8f644f5.png)

**Reply by borkdude** [reply]
And btw, you can also disable the warning one-time with `#_:clj-kondo/ignore`:

![Screenshot 2022-09-08 at 15 16 12](https://user-images.githubusercontent.com/284934/189131729-786fffbd-cb7a-4735-9733-2e228e5b4ac4.png)

**Reply by borkdude** [reply]
clj-kondo on the playground is now updated

**Comment by martianhi** [comment]
thanks for the extra info.

what i meant about errors is that i would get errors from the linter from any functions that were under the namespace that had been required with ` :refer :all`, as they are written with no namespace prefix at all.

but you're right, in the playground it works:
`[clojure.string :refer :all]`
gives a warning, then 
`(join "a" "b")` works with out linter complaints.

it's all good, i'm just getting the hang of my environment, takes a while...

thanks for sharing the details too.
